cpp/src/io/comp/nvcomp_adapter.cpp

```diff
@@ -321,40 +323,66 @@ static void batched_compress_async(compression_type compression,
   CUDF_EXPECTS(nvcomp_status == nvcompStatus_t::nvcompSuccess, "Error in compression");
 }
 
-bool is_aligned(void const* ptr, std::uintptr_t alignment) noexcept
+constexpr bool is_aligned(void const* ptr, std::uintptr_t alignment) noexcept
 {
   return (reinterpret_cast<std::uintptr_t>(ptr) % alignment) == 0;
 }
 
+// since some compressors have extreme temp memory requirements, scale back the number
+// of chunks to process until the request for temp memory succeeds.
+std::pair<rmm::device_buffer, size_t> compress_temp_buffer(compression_type compression,
+                                                           size_t num_chunks,
+                                                           size_t max_uncomp_chunk_size,
+                                                           rmm::cuda_stream_view stream)
+{
+  while (num_chunks > 0) {
+    try {
+      auto const temp_size =
+        batched_compress_temp_size(compression, num_chunks, max_uncomp_chunk_size);
+      rmm::device_buffer buf(temp_size, stream);
+      return std::pair(std::move(buf), num_chunks);
+    } catch (rmm::bad_alloc& ba) {
+      // don't loop forever...if num_chunks is already 1, the following divide will also
+      // yield 1
+      if (num_chunks == 1) { break; }
+      num_chunks = (num_chunks + 1) / 2;
```

Comment:
@vuule rather than halving each time, do you think it would be better to divide num_chunks by an increasing factor?  So for one example, 3 15GB passes rather than 4 12.5GB passes.

Reply:
I think this mostly depends on the performance of failed allocations. We're trading off more failed allocations for larger batch size.
Followed up on this for rmm, and learned that rmm performance in this case depends on the fragmentation, as it need to iterate over all free blocks before finding there are no blocks that are large enough.
That said, I would guess that it's still faster to increase the factor linearly, as you suggest.

Reply:
Is there a way to find the largest possible allocation in advance?  Then we could tune num_chunks to match that without actually trying the `allocate()` call.

Reply:
Good point. Yes, it would be better to throw `bad_alloc` so caller can programmatically get the reason for failure and act accordingly (e.g. spilling).

Reply:
Modifed to re-throw the `bad_alloc` when giving up.